src/redfish.c

```diff
@@ -0,0 +1,891 @@
+          ERROR(PLUGIN_NAME ": Type is not defined in property \"%s\" in "
+                            "query \"%s\"",
+                prop->name, query->name);
+          goto error;
+        }
+      }
+    }
+  }
+
+  c_avl_iterator_destroy(queries_iter);
+
+  return 0;
+
+error:
+  c_avl_iterator_destroy(queries_iter);
+  return -EINVAL;
+}
+
+static int redfish_convert_val(redfish_value_t *value,
+                               redfish_value_type_t src_type, value_t *vl,
+                               int dst_type) {
+  switch (dst_type) {
+  case DS_TYPE_GAUGE:
+    if (src_type == VAL_TYPE_STR)
+      vl->gauge = strtod(value->string, NULL);
+    else if (src_type == VAL_TYPE_INT)
+      vl->gauge = (gauge_t)value->integer;
+    else if (src_type == VAL_TYPE_REAL)
+      vl->gauge = value->real;
+    break;
+  case DS_TYPE_DERIVE:
+    if (src_type == VAL_TYPE_STR)
+      vl->derive = strtoll(value->string, NULL, 0);
+    else if (src_type == VAL_TYPE_INT)
+      vl->derive = (derive_t)value->integer;
+    else if (src_type == VAL_TYPE_REAL)
+      vl->derive = (derive_t)value->real;
+    break;
+  case DS_TYPE_COUNTER:
+    if (src_type == VAL_TYPE_STR)
+      vl->derive = strtoull(value->string, NULL, 0);
+    else if (src_type == VAL_TYPE_INT)
+      vl->derive = (derive_t)value->integer;
+    else if (src_type == VAL_TYPE_REAL)
+      vl->derive = (derive_t)value->real;
+    break;
+  case DS_TYPE_ABSOLUTE:
+    if (src_type == VAL_TYPE_STR)
+      vl->absolute = strtoull(value->string, NULL, 0);
+    else if (src_type == VAL_TYPE_INT)
+      vl->absolute = (absolute_t)value->integer;
+    else if (src_type == VAL_TYPE_REAL)
+      vl->absolute = (absolute_t)value->real;
+    break;
+  default:
+    ERROR(PLUGIN_NAME ": Invalid data set type. Cannot convert value");
+    return -EINVAL;
+  }
+
+  return 0;
+}
+
+static void redfish_process_payload(bool success, unsigned short http_code,
+                                    redfishPayload *payload, void *context) {
+  if (success == false) {
+    WARNING(PLUGIN_NAME ": Query has failed, HTTP code = %u\n", http_code);
+    return;
+  }
+  redfish_payload_ctx_t *res_serv = (redfish_payload_ctx_t *)context;
+  redfish_service_t *serv = res_serv->service;
+  if (payload) {
+    json_t *json_array;
+    for (llentry_t *llres = llist_head(res_serv->resources); llres != NULL;
+         llres = llres->next) {
+      redfish_resource_t *res = (redfish_resource_t *)llres->value;
+      json_array = json_object_get(payload->json, res->name);
+      for (llentry_t *llprop = llist_head(res->properties); llprop != NULL;
+           llprop = llprop->next) {
+        redfish_property_t *prop = (redfish_property_t *)llprop->value;
+        if (json_array == NULL) {
+          ERROR("Could not find resource \"%s\"", res->name);
+          continue;
+        }
+
+        /* Iterating through array of sensor(s) */
+        for (int i = 0; i < json_array_size(json_array); i++) {
```

Comment:
This is the third nested loop. Please break this function up into smaller functiosn that are easier to comprehend.